app/models/receptacle/downstream_aliquots_removal.rb

```diff
@@ -0,0 +1,44 @@
+# frozen_string_literal: true
+
+# DPL-451: Added functonality to not remove aliquots when failing plates that have a sequencing batch
+# downstream
+module Receptacle::DownstreamAliquotsRemoval
+  # Public interface for the downstream aliquots removal functionality
+  module Mixin
+    # Returns a boolean that indicates if it is valid to remove downstream aliquots starting from this one.
+    # Current condition (DPL-451) estimates that if the labware has descendants that are involved
+    # in a sequencing batch, we do not remove any data from downstream aliquots or from MLWH.
+    #
+    # @return [Boolean] true if we allow to remove downstream aliquots, false otherwise
+    def allow_to_remove_downstream_aliquots?
+      creation_batches = PrivateMethods.creation_batches_for_requests(self)
+      creation_batches.nil? || creation_batches.flatten.uniq.length.zero?
+    end
+  end
+
+  # Private methods to provide functionality for the mixin
+  module PrivateMethods
+    # Gets the submissions of all outer requests that wrap this well
+    #
+    # @return [Array[Submission]] List of submissions
+    def self.submissions_for_requests(instance)
+      instance.aliquot_requests.map(&:submission)&.flatten&.uniq
+    end
+
+    # Get the sequencing creations batches by following the outer requests graph.
+    # Supposing this outer requests graph:
+    # Library Creation ---> Multiplexing --> Sequencing
+    # The path to solve the creation batches is as follows:
+    # 1) Get the submission from the outer requests that wrap this current node
+    # 2) From the first submission, get the multiplexed labware, which is the labware
+    #    that is the target of a multiplexed request (outer request before sequencing)
+    # 3) From this multiplexed labware, it goes to the next labware created from it (sequencing tube).
+    # 4) For this labware, it access to the creation batches (groups of sequencing requests that will go into
+    #    a flowcell)
+    #
+    # @return [Array[Batch]] List of batches
+    def self.creation_batches_for_requests(instance)
+      PrivateMethods.submissions_for_requests(instance)&.first&.multiplexed_labware&.children&.map(&:creation_batches)
```

Comment:
`PrivateMethods.submissions_for_requests(instance)&.first&.multiplexed_labware`
... will this work if the submission we're looking at is a library-only submission, and the sequencing request is made separately, later (e.g. using the 'request additional sequencing' button)?

Reply:
Yes, I suppose not. I'll modify it so it check all submissions.